pkg/scheduler/framework/plugins/nodevolumelimits/csi.go

```diff
@@ -227,6 +235,10 @@ func (pl *CSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod *v
 		maxVolumeLimit, ok := nodeVolumeLimits[driverName]
 		if ok {
 			currentVolumeCount := attachedVolumeCount[driverName]
+			volumeAttachmentsCount := volumeAttachments[driverName]
+			if volumeAttachmentsCount > currentVolumeCount {
+				currentVolumeCount = volumeAttachmentsCount
+			}
```

Comment:
I am afraid the whole decision if a pod is schedulable is far more complex.

In `attachedVolumes` above, the scheduler has all volumes of all pods that are scheduled to that node. It does not include volumes of pods that were already deleted, but their VolumeAttachment still exists and the volumes are attached.

`volumeAttachments` contains volumes that are _currently_ attached. It does include volumes of deleted pods, on the other hand, it does not include volumes of pods that were scheduled to the node, but VolumeAttachment does not exist yet.

Evaluating each of these lists separately will produce wrong numbers. We need to merge these two lists by volumeID (`csiDriverNameName + '/' + volumeHandle`) - some volumes will be in both (pod is scheduled + VolumeAttachment exists), some will be only in `attachedVolumes` (pod is scheduled, but VolumeAttachment does not exist yet) and some will be only in `volumeAttachments` (pod got deleted, but VolumeAttachment still exists).

Reply:
That makes sense, thank you Jan for the insightful review. This is fixed in latest rev.